packages/expo-device/ios/EXDevice/EXDeviceUID.m

```diff
@@ -0,0 +1,144 @@
+ At last, the UID is persisted if needed to.
+ */
+- (NSString *)uid {
+  if (!_uid) _uid = [[self class] valueForKeychainKey:_uidKey service:_uidKey];
+  if (!_uid) _uid = [[self class] valueForUserDefaultsKey:_uidKey];
+  if (!_uid) _uid = [[self class] appleIFV];
+  if (!_uid) _uid = [[self class] randomUUID];
+  [self save];
+  return _uid;
+}
+
+/*! Persist UID to NSUserDefaults and Keychain, if not yet saved
+ */
+- (void)save {
+  if (![EXDeviceUID valueForUserDefaultsKey:_uidKey]) {
+    [EXDeviceUID setValue:_uid forUserDefaultsKey:_uidKey];
+  }
+  if (![EXDeviceUID valueForKeychainKey:_uidKey service:_uidKey]) {
+    [EXDeviceUID setValue:_uid forKeychainKey:_uidKey inService:_uidKey];
+  }
+}
+
+#pragma mark - Keychain methods
+
+/*! Create as generic NSDictionary to be used to query and update Keychain items.
+ *  param1
+ *  param2
+ */
++ (NSMutableDictionary *)keychainItemForKey:(NSString *)key service:(NSString *)service {
+  NSMutableDictionary *keychainItem = [[NSMutableDictionary alloc] init];
+  keychainItem[(__bridge id)kSecClass] = (__bridge id)kSecClassGenericPassword;
+  keychainItem[(__bridge id)kSecAttrAccessible] = (__bridge id)kSecAttrAccessibleAlways;
+  keychainItem[(__bridge id)kSecAttrAccount] = key;
+  keychainItem[(__bridge id)kSecAttrService] = service;
+  return keychainItem;
+}
+
+/*! Sets
+ *  param1
+ *  param2
+ */
++ (OSStatus)setValue:(NSString *)value forKeychainKey:(NSString *)key inService:(NSString *)service {
+  NSMutableDictionary *keychainItem = [[self class] keychainItemForKey:key service:service];
+  keychainItem[(__bridge id)kSecValueData] = [value dataUsingEncoding:NSUTF8StringEncoding];
+  return SecItemAdd((__bridge CFDictionaryRef)keychainItem, NULL);
+}
+
++ (NSString *)valueForKeychainKey:(NSString *)key service:(NSString *)service {
+  OSStatus status;
+  NSMutableDictionary *keychainItem = [[self class] keychainItemForKey:key service:service];
+  keychainItem[(__bridge id)kSecReturnData] = (__bridge id)kCFBooleanTrue;
+  keychainItem[(__bridge id)kSecReturnAttributes] = (__bridge id)kCFBooleanTrue;
+  CFDictionaryRef result = nil;
+  status = SecItemCopyMatching((__bridge CFDictionaryRef)keychainItem, (CFTypeRef *)&result);
+  if (status != noErr) {
+    return nil;
+  }
+  NSDictionary *resultDict = (__bridge_transfer NSDictionary *)result;
+  NSData *data = resultDict[(__bridge id)kSecValueData];
+  if (!data) {
+    return nil;
+  }
+  return [[NSString alloc] initWithData:data encoding:NSUTF8StringEncoding];
+}
+
+#pragma mark - NSUserDefaults methods
+
++ (BOOL)setValue:(NSString *)value forUserDefaultsKey:(NSString *)key {
+  [[NSUserDefaults standardUserDefaults] setObject:value forKey:key];
+  return [[NSUserDefaults standardUserDefaults] synchronize];
+}
+
++ (NSString *)valueForUserDefaultsKey:(NSString *)key {
+  return [[NSUserDefaults standardUserDefaults] objectForKey:key];
+}
+
+#pragma mark - UID Generation methods
+
++ (NSString *)appleIFV {
```

Comment:
I'm fairly sure Apple restricts the use of this ID pretty heavily, and I'm not sure we want to include it here because it can cause apps (even apps that aren't using it in JS) to be flagged during Apple review.

Reply:
I think this ID is OK actually -- it's not the IDFA (nor UDID but that is no longer exposed except through the provisioning profile dance). But maybe it does make sense to compartmentalize the fields we think are riskier (I'm suspect of the the CoreTelephony ones, though haven't done any research to see if they cause issues).